src/Data/Array/Mutable/Linear.hs

```diff
@@ -167,18 +168,40 @@ resize newSize seed (Array _ mut)
   | otherwise =
       Array newSize (Unsafe.resizeMutArr mut seed newSize)
 
--- XXX: Replace with toVec
+-- | Return the array elements as a spine-strict list. This
+-- function is expensive since it has to traverse the entire
+-- array before returning.
 toList :: Array a #-> (Array a, Unrestricted [a])
-toList arr = size arr & \case
-  (arr', len) -> move len & \case
-    Unrestricted len' -> toListWalk (len' - 1) arr' (Unrestricted [])
+toList arr =
+  size arr & \(arr', Unrestricted len) ->
+    toListWalk (len - 1) arr' (Unrestricted [])
   where
   toListWalk :: Int -> Array a #-> Unrestricted [a] -> (Array a, Unrestricted [a])
   toListWalk ix arr accum
     | ix < 0 = (arr, accum)
     | otherwise = read arr ix & \case
         (arr', Unrestricted x) -> toListWalk (ix - 1) arr' ((x:) Control.<$> accum)
 
+-- We implement 'toListLazy' below relying on two things:
+--
+-- * We hold the only reference to the underlying 'MutArr#'.
+-- Otherwise someone else could write to it before the returned
+-- list is evaluated.
+-- * 'Consume' instance of 'Array' is no-op.
+
+-- | Return the array elements as a lazy list. The array
+-- lookups will happen lazily, so this function also consumes
+-- the array.
```

Comment:
You probably want to explain this a little better.

Also, maybe this ought to be the default, and the other one the one with a long name. I strongly suspect that this is a more common case for `toList`: you do `toList` to have a lazy traversal of your array, and probably you are abandoning the array there. (the other one is “just” the composition of `dup2` and `toList` (which, by the way, is easier to express with the `_2` lens, maybe that's how it ought to be defined (when the `Dupable` instance is implemented)))

Reply:
> You probably want to explain this a little better.

Agreed, will do.

> ... the other one is “just” the composition of dup2 and toList ....

This is a great point, thank you; I completely agree. So, now my plan is to drop the strict function, make the lazy function the default. Once we have the dupable instance (awaiting #151), it'd be trivial to implement that.

I will do the same to the `Vector` interface too.